### PR TITLE
Upgrade to Ocaml 413

### PR DESCRIPTION
**Pull request by voodoos**
@let-def I merged your changes to synchronize with 4.13.1

Did you also start reviewing the upgrade ? This new PR is the right place to make comments.

### CODE REVIEW

**Comment by voodoos** [comment]
@let-def , @trefis : any intuition regarding the CI failure ?

Only happening on `debian-11-4.13_x86_32` :

```diff
--- a/../../../default/tests/test-dirs/document/src-documentation.t/run.t
+++ b/tests/test-dirs/document/src-documentation.t/run.t.corrected
@@ -43,6 +43,9 @@ documentation for the non-last defined value (in the same file) is show
   " List reversal. "
 
   $ dune build --root=. ./doc.exe
+      ocamlopt doc.exe
+  /usr/bin/ld: .doc.eobjs/native/doc.o: warning: relocation in read-only section `.text'
+  /usr/bin/ld: warning: creating DT_TEXTREL in a PIE
   $ cat >.merlin <<EOF
   > B _build/default/.doc.eobjs/byte
   > S .
   ```

**Comment by nojb** [comment]
> @let-def , @trefis : any intuition regarding the CI failure ?
> 
> Only happening on `debian-11-4.13_x86_32` :
> 
> ```diff
> --- a/../../../default/tests/test-dirs/document/src-documentation.t/run.t
> +++ b/tests/test-dirs/document/src-documentation.t/run.t.corrected
> @@ -43,6 +43,9 @@ documentation for the non-last defined value (in the same file) is show
>    " List reversal. "
>  
>    $ dune build --root=. ./doc.exe
> +      ocamlopt doc.exe
> +  /usr/bin/ld: .doc.eobjs/native/doc.o: warning: relocation in read-only section `.text'
> +  /usr/bin/ld: warning: creating DT_TEXTREL in a PIE
> ```

This is a 32-bit system, right? See https://github.com/ocaml/ocaml/issues/9800

**Comment by nojb** [comment]
> > @let-def , @trefis : any intuition regarding the CI failure ?
> This is a 32-bit system, right? See [ocaml/ocaml#9800](https://github.com/ocaml/ocaml/issues/9800)

**Comment by voodoos** [comment]
Oh right, that must be it !
Thanks for the pointer

**Comment by let-def** [comment]
I will review again today, but it already looked good last time I looked.

**Comment by voodoos** [comment]
> I will review again today, but it already looked good last time I looked.

Thank you !

For the failing test I guess we can just disable it with `(enabled_if (<> %{architecture} "86_32"))`

File: src/analysis/construct.ml [hunk 1]
```diff
@@ -441,8 +443,8 @@ module Gen = struct
           with Not_found ->
             let def = Env.find_type_descrs path env in
             match def with
-            | constrs, [] -> constructor env rtyp path constrs
-            | [], labels -> record env rtyp path labels
+            | Type_variant (constrs, _) -> constructor env rtyp path constrs
+            | Type_record (labels, _) -> record env rtyp path labels
             | _ -> []
```
Comment (let-def) [inline src/analysis/construct.ml:448]: `| Type_abstract | Type_open -> []`
Now that there is a proper representation of all cases, better match on them :).

File: src/analysis/destruct.ml [hunk 2]
```diff
@@ -197,8 +194,12 @@ let rec gen_patterns ?(recurse=true) env type_expr =
               Patterns.omegas cstr_descr.cstr_arity
           in
           let lidl = Location.mknoloc (prefix cstr_descr.cstr_name) in
-          Some (Tast_helper.Pat.construct env type_expr lidl cstr_descr args)
+          Some (
+            Tast_helper.Pat.construct env type_expr lidl cstr_descr args None)
       )
+    | _ ->
+      if recurse then from_type_decl env path type_expr else
+      raise (Not_allowed (sprintf "non-destructible type: %s" (Path.last path)))
```
Comment (let-def) [inline src/analysis/destruct.ml:202]: Same here, match `Type_abstract | Type_open`

File: src/analysis/destruct.ml [hunk 3]
```diff
@@ -594,16 +597,18 @@ let rec node config source selected_node parents =
...
             | [ more_precise ] ->
+              (* Printf.eprintf "one cp \n%!" ; *)
               (* If only one pattern is generated, then we're only refining the
                 current pattern, not generating new branches. *)
               let ppat = filter_pat_attr (Untypeast.untype_pattern more_precise) in
               let str = Mreader.print_pretty
                   config source (Pretty_pattern ppat) in
               patt.Typedtree.pat_loc, str
-            | sub_patterns ->
+            | h::_ as sub_patterns ->
```
Comment (let-def) [inline src/analysis/destruct.ml:611]: `h` does not seem to be used... Why adding this particular pattern?

File: src/ocaml/parsing/pprintast.ml [hunk 4]
```diff
@@ -1677,8 +1709,9 @@ let core_type = core_type reset_ctxt
 let pattern = pattern reset_ctxt
 let signature = signature reset_ctxt
 let structure = structure reset_ctxt
+let module_expr = module_expr reset_ctxt
 let case_list = case_list reset_ctxt
-let module_ = module_expr reset_ctxt
+let module_ = module_expr (* for merlin *)
```
Comment (let-def) [inline src/ocaml/parsing/pprintast.ml:1714]: `module_ = module_expr`...
Maybe we can remove `module_` (added for merlin), now that `module_expr` is exported and used by the upstream typechecker (for printing diffs it seems)?

**Review by let-def — APPROVED** [review]
As far as I am concerned, the PR could be merged as is.
I suggest a few cosmetic changes that could simplify maintenance, but nothing important.

**Comment by nojb** [comment]
Friendly ping. Is there anything left to do here? It would be nice to have an official release for 4.13. Thanks!

**Comment by voodoos** [comment]
> Friendly ping. Is there anything left to do here? It would be nice to have an official release for 4.13. Thanks!

We plan to submit the PR to opam repository before the end of the week.

**Pull request merged**